training/creating_dataset.py

```diff
@@ -42,13 +83,149 @@ def is_path_correct(path: str) -> bool:
     """
     if not os.path.isfile(path):
         raise ValueError(f"Path {path} does not exist")
-        # return False
+
     if not (path.endswith(".db") or path.endswith(".json")):
         raise ValueError(f"Expected db or json file, got {path}")
-        # return False
+
     return True
 
 
+def split_characteristics_into_groups(
+    characteristics: list,
+) -> tuple[set[str], set[str], set[str], set[str], list[str]]:
+    """
+    Split given characteristics list into five categories according to sets and dictionaries with characteristics.
+    Categories depend on way of extracting data.
+    :param characteristics: list of characteristics
+    :return: sets and list of characteristics after split
+    """
+    speeds = OTHER_SPEEDS.intersection(characteristics)
+    weaknesses = WEAKNESSES.intersection(characteristics)
+    resistances = RESISTANCES.intersection(characteristics)
+    special_characteristics = SPECIAL_CHARACTERISTICS.intersection(characteristics)
+    characteristics_rename = [
+        ch for ch in characteristics if ch in CHARACTERISTICS_RENAME.keys()
+    ]
+
+    return (
+        speeds,
+        weaknesses,
+        resistances,
+        special_characteristics,
+        characteristics_rename,
+    )
+
+
+def get_characteristic_from_list(cell_value, characteristic_name: str) -> int:
+    """
+    Function used for pd.Series.apply()\n
+    Retrieve the value associated with a specific characteristic from a list of dictionaries.
+    :param cell_value: Value of cell from df. A list of dictionaries, where each dictionary represents a characteristic
+            with 'type' and 'value' keys.
+    :param characteristic_name: The characteristic name to search for.
+    :return: The value associated with the specified characteristic if found, or 0 if not found.
+    """
+    if not cell_value or np.all(pd.isnull(cell_value)):
+        return 0
+
+    result = [
+        x.get("value") for x in cell_value if x.get("type") == characteristic_name
+    ]
+    return 0 if len(result) == 0 else result[0]
+
+
+def get_nr_of_spells_with_lvl(items_list: list[dict], spell_level: int) -> int:
+    """
+    Function used for pd.Series.apply()\n
+    Count the number of spells in the given list that match the specified level.
+    :param items_list: A list of dictionaries representing items, each with relevant attributes.
+    :param spell_level: The level to filter spells by.
+    :return: The count of spells in the list that match the specified level.
+    """
+    # get only spell items
+    spells = [
+        i
+        for i in items_list
+        if i.get("type") == "spell"
+        and i.get("system").get("category").get("value") == "spell"
+        # skip cantrip spells
+        and "cantrip" not in i.get("system").get("traits").get("value")
+        and i.get("system").get("level").get("value") == spell_level
```

Comment:
Jeżeli już wiemy, że to są czary (inaczej pierwszy warunek w `if` już by to ominął), to nie ma gwarancji, że istnieją klucze `system` i `category` przynajmniej? Bo w Pythonie if ma short-circuting, tzn. jak wcześniejszy warunek nie jest spełniony, to w ogóle nie przejdzie do kolejnych, więc być może mamy zagwarantowane, że te klucze istnieją, jak w ogóle możemy dojść do tych dalszych warunków.

Reply:
Chyba nie do końca rozumiem 😅 

Jeżeli to są czary to `system` i ` category` istnieją - przynajmniej w tych 3 bestiariuszach
Z tego co widziałam to `system` chyba zawsze istnieje

Wszystkie warunki dałam tak, żeby uzyskać tylko te "czary czary" o które chodzi:
` i.get("system").get("category").get("value") == "spell"` - eliminacja np. "ritual"
`"cantrip" not in i.get("system").get("traits").get("value")` - eliminacja cantrip
`i.get("system").get("level").get("value") == spell_level` - odpowiedni level

Reply:
W sensie chodzi mi o ograniczenie ilości get'ów, bo są długie. Przykładowo:
```
i.get("system").get("category").get("value") == "spell"
->
i["system"]["category"]["value"]
```
W sensie moim zdaniem jest jednak czytelniej z bracketami